<?xml version="1.0"?>
<!-- Revolute-Revolute Manipulator -->
<robot name="ugv" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <!-- Constants for robot dimensions -->
  <xacro:property name="PI" value="3.1415926535897931"/>
  <xacro:property name="mass" value="5" />
  <xacro:property name="length" value="0.813" />
  <xacro:property name="width" value="0.5045" />
  <xacro:property name="height" value="0.202" />
  <xacro:property name="cylinder_length" value="0.025" />
  <xacro:property name="cylinder_radius" value="0.055" />
  <xacro:property name="width_diff" value="0.05" />
  <xacro:property name="height_diff" value="0.05" />
  <xacro:property name="length_diff" value="0.05" />

  <!-- Import all Gazebo-customization elements, including Gazebo colors -->
  <!-- <xacro:include filename="$(find testing)/urdf/ugv.gazebo" /> -->
  <!-- Import Rviz colors -->
  <!-- <xacro:include filename="$(find testing)/urdf/materials.xacro" /> -->

  <!-- Used for fixing robot to Gazebo 'base_link' -->
  <link name="odom"/>

  <joint name="footprint_joint" type="fixed">
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <parent link="odom"/>
    <child link="chassis"/>
  </joint>


  <!-- Base Link -->
  <link name="chassis">
    <collision>
      <origin xyz="0 0 ${height/2.0}" rpy="0 0 0"/>
      <geometry>
	<box size="${width} ${width} ${height}"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 ${height/2.0}" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://testing/meshes/UGV.stl"/>
      </geometry>
    </visual>

    <inertial>
      <origin xyz="0 0 ${height/2.0}" rpy="0 0 0"/>
      <mass value="${mass}"/>
      <inertia
	  ixx="${mass / 12.0 * (width*width + height*height)}" ixy="0.0" ixz="0.0"
	  iyy="${mass / 12.0 * (height*height + width*width)}" iyz="0.0"
	  izz="${mass / 12.0 * (width*width + width*width)}"/>
    </inertial>
  </link>  

  <!-- HOKUYO LASER -->
  <joint name="hokuyo_joint" type="fixed">
    <axis xyz="0 0 0" />
    <origin xyz="0.27 0 ${height + 0.025}" rpy="0 0 0"/>
    <parent link="chassis"/>
    <child link="hokuyo_link"/>
  </joint>
  
  <!-- Hokuyo Laser -->
  <link name="hokuyo_link">
    <collision>
      <origin xyz="0 0 -0.025" rpy="0 0 0"/>
      <geometry>
	<box size="0.051 0.051 0.07"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 -0.05" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://testing/meshes/Hokuyo.stl"/>
      </geometry>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 -0.05" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

  <!-- CAMERA -->
  <joint name="camera_joint" type="fixed">
    <axis xyz="0 0 0" />
    <origin xyz="0.3 0.1 ${height}" rpy="0 0 0"/>
    <parent link="chassis"/>
    <child link="camera_link"/>
  </joint>

  <!-- Camera -->
  <link name="camera_link">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
  	<box size="0.05 0.05 0.05"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
  	<box size="0.05 0.05 0.05"/>
      </geometry>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

  <!-- IMU -->
  <joint name="imu_joint" type="fixed">
    <axis xyz="0 0 0" />
    <origin xyz="0 0 ${height}" rpy="0 0 0"/>
    <parent link="chassis"/>
    <child link="imu_link"/>
  </joint>

  <link name="imu_link">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
  	<box size="0.05 0.05 0.05"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
  	<box size="0.05 0.05 0.05"/>
      </geometry>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>
  
</robot>
